In [3]:
from ultralytics import YOLO

# 1. Load the pre-trained YOLOv8 Nano base model
model = YOLO("yolov8n.pt")

# 2. Train on the 5,000-image dataset
model.train(
    data="dataset.yaml",
    epochs=15,           # 10 to 20 epochs gives good convergence
    batch=32,            # Use batch=32 (or 16 if training on CPU / lower VRAM)
    optimizer="Adam",
    lrf=0.001,
    imgsz=640
)


New https://pypi.org/project/ultralytics/8.4.173 available  Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.2.73  Python-3.11.3 torch-2.4.0+cpu CPU (AMD Ryzen 5 4600H with Radeon Graphics)
engine\trainer: task=detect, mode=train, model=yolov8n.pt, data=dataset.yaml, epochs=15, time=None, patience=100, batch=32, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train3, exist_ok=False, pretrained=True, optimizer=Adam, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_fr

train: Scanning C:\Users\tanis\Documents\DS MPR\Real_time_drowsy_driving_detection\datasets\labels\train... 5000 images, 0 backgrounds, 0 corrupt: 100%|██████████| 5000/5000 [00:15<00:00, 326.22it/s]


train: New cache created: C:\Users\tanis\Documents\DS MPR\Real_time_drowsy_driving_detection\datasets\labels\train.cache


val: Scanning C:\Users\tanis\Documents\DS MPR\Real_time_drowsy_driving_detection\datasets\labels\val... 5008 images, 0 backgrounds, 0 corrupt: 100%|██████████| 5008/5008 [00:15<00:00, 318.52it/s]


val: New cache created: C:\Users\tanis\Documents\DS MPR\Real_time_drowsy_driving_detection\datasets\labels\val.cache
Plotting labels to C:\Users\tanis\Documents\DS MPR\Real_time_drowsy_driving_detection\runs\detect\train3\labels.jpg... 
optimizer: Adam(lr=0.01, momentum=0.937) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 0 dataloader workers
Logging results to C:\Users\tanis\Documents\DS MPR\Real_time_drowsy_driving_detection\runs\detect\train3
Starting training for 15 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/15         0G        nan        nan        nan        103        640:   3%|▎         | 4/157 [04:55<3:08:37, 73.97s/it] 


KeyboardInterrupt: 

New https://pypi.org/project/ultralytics/8.2.75 available  Update with 'pip install -U ultralytics'
Ultralytics YOLOv8.2.71  Python-3.9.13 torch-1.13.1+cu117 CUDA:0 (NVIDIA GeForce RTX 4070 Ti, 12281MiB)
WARNING  Upgrade to torch>=2.0.0 for deterministic training.
engine\trainer: task=detect, mode=train, model=yolov8n.pt, data=dataset.yaml, epochs=10, time=None, patience=100, batch=64, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train, exist_ok=False, pretrained=True, optimizer=Adam, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.001, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, bgr=0.0, mosaic=1.0, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=C:\Utils\Real_time_drowsy_driving_detection\runs\detect\train
Overriding model.yaml nc=80 with nc=2

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]             
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]             
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]               
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]           
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]              
  8                  -1  1    460288  ultralytics.nn.modules.block.C2f             [256, 256, 1, True]           
  9                  -1  1    164608  ultralytics.nn.modules.block.SPPF            [256, 256, 5]                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 12                  -1  1    148224  ultralytics.nn.modules.block.C2f             [384, 128, 1]                 
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 15                  -1  1     37248  ultralytics.nn.modules.block.C2f             [192, 64, 1]                  
 16                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
 17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 18                  -1  1    123648  ultralytics.nn.modules.block.C2f             [192, 128, 1]                 
 19                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 21                  -1  1    493056  ultralytics.nn.modules.block.C2f             [384, 256, 1]                 
 22        [15, 18, 21]  1    751702  ultralytics.nn.modules.head.Detect           [2, [64, 128, 256]]           
Model summary: 225 layers, 3,011,238 parameters, 3,011,222 gradients, 8.2 GFLOPs

Transferred 319/355 items from pretrained weights
Freezing layer 'model.22.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed 
train: Scanning C:\Utils\Real_time_drowsy_driving_detection\datasets\labels\train... 53056 images, 0 backgrounds, 0 corrupt: 100%|██████████| 53056/53056 [00:39<00:00, 1356.03it/s]
train: New cache created: C:\Utils\Real_time_drowsy_driving_detection\datasets\labels\train.cache
val: Scanning C:\Utils\Real_time_drowsy_driving_detection\datasets\labels\val... 2535 images, 0 backgrounds, 0 corrupt: 100%|██████████| 2535/2535 [00:02<00:00, 879.95it/s] 
val: New cache created: C:\Utils\Real_time_drowsy_driving_detection\datasets\labels\val.cache
Plotting labels to C:\Utils\Real_time_drowsy_driving_detection\runs\detect\train\labels.jpg... 
optimizer: Adam(lr=0.01, momentum=0.937) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 8 dataloader workers
Logging results to C:\Utils\Real_time_drowsy_driving_detection\runs\detect\train
Starting training for 10 epochs...
Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       1/10      8.54G     0.8405       1.23      1.446         64        640: 100%|██████████| 829/829 [04:36<00:00,  3.00it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.79it/s]                   all       2535       2535      0.461      0.457      0.505      0.483


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       2/10      8.54G     0.7491      1.006      1.386         64        640: 100%|██████████| 829/829 [04:32<00:00,  3.04it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.83it/s]                   all       2535       2535      0.531      0.498      0.412      0.409


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       3/10      8.53G     0.6953     0.9423      1.334         64        640: 100%|██████████| 829/829 [04:31<00:00,  3.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.86it/s]                   all       2535       2535      0.503      0.553      0.519      0.389


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       4/10      8.53G     0.6362     0.8711       1.28         64        640: 100%|██████████| 829/829 [04:31<00:00,  3.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.83it/s]                   all       2535       2535      0.508      0.492      0.482      0.478


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       5/10      8.53G     0.5895     0.8148      1.233         64        640: 100%|██████████| 829/829 [04:31<00:00,  3.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.85it/s]                   all       2535       2535      0.922      0.497      0.518      0.475


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       6/10      8.53G     0.5553     0.7772      1.201         64        640: 100%|██████████| 829/829 [04:33<00:00,  3.03it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:07<00:00,  2.85it/s]                   all       2535       2535      0.421      0.569      0.526      0.502


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       7/10      8.53G     0.5289     0.7391      1.177         64        640: 100%|██████████| 829/829 [04:31<00:00,  3.05it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.86it/s]                   all       2535       2535      0.634      0.706      0.652      0.619


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       8/10      8.53G      0.501      0.705      1.151         64        640: 100%|██████████| 829/829 [04:30<00:00,  3.06it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.86it/s]                   all       2535       2535      0.661      0.795      0.739      0.668


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       9/10      8.53G     0.4741     0.6708      1.132         64        640: 100%|██████████| 829/829 [04:31<00:00,  3.06it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.87it/s]                   all       2535       2535      0.762      0.791      0.782       0.72


      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      10/10      8.53G      0.446     0.6383      1.109         64        640: 100%|██████████| 829/829 [04:30<00:00,  3.06it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:06<00:00,  2.87it/s]                   all       2535       2535       0.73       0.86      0.784      0.732


10 epochs completed in 0.782 hours.
Optimizer stripped from C:\Utils\Real_time_drowsy_driving_detection\runs\detect\train\weights\last.pt, 6.2MB
Optimizer stripped from C:\Utils\Real_time_drowsy_driving_detection\runs\detect\train\weights\best.pt, 6.2MB

Validating C:\Utils\Real_time_drowsy_driving_detection\runs\detect\train\weights\best.pt...
Ultralytics YOLOv8.2.71  Python-3.9.13 torch-1.13.1+cu117 CUDA:0 (NVIDIA GeForce RTX 4070 Ti, 12281MiB)
Model summary (fused): 168 layers, 3,006,038 parameters, 0 gradients, 8.1 GFLOPs
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 20/20 [00:08<00:00,  2.45it/s]
                   all       2535       2535      0.731      0.859      0.784      0.733
              open eye        969        969      0.514      0.729      0.577      0.474
             close eye       1566       1566      0.948      0.989      0.991      0.991
Speed: 0.1ms preprocess, 1.2ms inference, 0.0ms loss, 0.4ms postprocess per image
Results saved to C:\Utils\Real_time_drowsy_driving_detection\runs\detect\train